src/timepicker/TimePickerState.ts

```diff
@@ -1,68 +1,106 @@
+
+  const defaultValue =
+    defaultValueProp && isValidTime(defaultValueProp)
+      ? parseTime(defaultValueProp)
+      : parseTime(`${new Date().getHours()}:${new Date().getMinutes()}`);
+  const time = timeProp == null ? defaultValue : parseTime(timeProp);
 
-  React.useEffect(() => {
-    onChange?.(time);
-  }, [onChange, time]);
+  const setTime = (hour: number, minute: number, meridian: string) => {
+    if (meridian === "PM") {
+      hour = +hour + 12;
+    } else {
+      hour = hour % 12;
+    }
+
+    setTimeProp(`${pad(hour)}:${pad(minute)}`);
+  };
+
+  const [hour, setHourProp] = React.useState(time?.getHours() % 12);
+  const [minute, setMinuteProp] = React.useState(time?.getMinutes());
+  const [meridian, setMeridianProp] = React.useState(
+    time?.getHours() >= 12 ? "PM" : "AM",
+  );
+
+  const setHour = (hour: number) => {
+    setHourProp(hour);
+    setTime(hour, minute, meridian);
+  };
+
+  const setMinute = (minute: number) => {
+    setMinuteProp(minute);
+    setTime(hour, minute, meridian);
+  };
+
+  const setMeridian = (meridianValue: string) => {
+    setMeridianProp(meridianValue);
+    setTime(hour, minute, meridianValue);
+  };
 
   const hourState = useTimePickerColumnState({
-    value: time!.getHours() % 12,
-    onChange: v => {
-      const d = new Date(time?.toISOString() as string);
-      d.setHours(v as number);
-      setTime(d);
-    },
-    onSelection: () => disclosure.hide(),
+    value: hour,
+    onChange: setHour,
+    onSelection: () => popover.hide(),
   });
 
   const minuteState = useTimePickerColumnState({
-    value: time?.getMinutes(),
-    onChange: v => {
-      const d = new Date(time?.toISOString() as string);
-      d.setMinutes(v as number);
-      setTime(d);
-    },
-    onSelection: () => disclosure.hide(),
+    value: minute,
+    onChange: setMinute,
+    onSelection: () => popover.hide(),
   });
 
   const meridiesState = useTimePickerColumnState({
-    value: time?.toLocaleTimeString().slice(-2).toUpperCase(),
-    onChange: v => {
-      const d = new Date(time?.toISOString() as string);
-      const hour = d.getHours() % 12;
-
-      if (v === "AM") {
-        d.setHours(hour);
-      } else {
-        d.setHours(hour + 12);
-      }
-      setTime(d);
-    },
-    onSelection: () => disclosure.hide(),
+    value: meridian,
+    onChange: setMeridian,
+    onSelection: () => popover.hide(),
   });
 
-  const composite = useCompositeState({ orientation: "horizontal" });
-
   const hours = [...new Array(13).keys()].slice(1);
   const minutes = [...new Array(60).keys()];
   const meridies = ["AM", "PM"];
+
+  const setSegmentTime = (date: Date) => {
+    setTimeProp(`${date.getHours()}:${date.getMinutes()}`);
+  };
+
+  const segmentState = useSegmentState({
+    value: time,
+    defaultValue,
+    onChange: setSegmentTime,
+    formatOptions,
+    placeholderDate,
+  });
+
+  const popover = usePopoverState({ modal: true, ...props });
```

Comment:
Move this to the beginning, since we are using popover before `onSelection: () => popover.hide(),` we are not getting error since it's wrapped in a function if we wanted to do `onSelection: popover.hide` it will throw error